protovalidate/internal/string_format.py

```diff
@@ -146,16 +146,30 @@ def format_string(self, arg: celtypes.Value) -> celpy.Result:
         if isinstance(arg, celtypes.StringType):
             return arg
         if isinstance(arg, celtypes.BytesType):
-            return celtypes.StringType(arg.hex())
+            return celtypes.StringType(arg)
         if isinstance(arg, celtypes.ListType):
             return self.format_list(arg)
+        if isinstance(arg, celtypes.BoolType):
+            # True -> true
+            return celtypes.StringType(str(arg).lower())
+        if isinstance(arg, celtypes.DoubleType):
+            return celtypes.StringType(f"{arg:.0f}")
+        if isinstance(arg, celtypes.TimestampType):
+            base = arg.isoformat()
+            if arg.getMilliseconds() != 0:
+                base = arg.isoformat(timespec="milliseconds")
+            return celtypes.StringType(base.removesuffix("+00:00") + "Z")
```

Comment:
just for context, _most_ protovalidate rules are defined in CEL using the `%s` verb (in fact, I don't think I've seen another verb used). `format_string` handles that verb, so we need to add some additional handling here.